clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -258,8 +258,9 @@ public synchronized int sendFetches() {
             if (log.isDebugEnabled()) {
                 log.debug("Sending {} {} to broker {}", isolationLevel, data.toString(), fetchTarget);
             }
-            client.send(fetchTarget, request)
-                    .addListener(new RequestFutureListener<ClientResponse>() {
+            RequestFuture<ClientResponse> future = client.send(fetchTarget, request);
+            this.nodesWithPendingFetchRequests.add(entry.getKey().id());
+            future.addListener(new RequestFutureListener<ClientResponse>() {
```

Comment:
Wondering if the subtleties of this fairly benign-looking code warrant a comment here or in the test?

Reply:
Was hoping the test would catch the issues with it - but I guess it's possible in the future some change causes the test to be less useful. Could stick a comment to the effect of "ensure `nodesWithPendingFetchRequests` is always present before finalizer could be called" or something.

Reply:
Was also wondering if there could ever be an exception thrown by `addListener` which would cause the listener to not be added or the completion handler to not be called?

Reply:
> Was also wondering if there could ever be an exception thrown by addListener which would cause the listener to not be added or the completion handler to not be called?

Hm good question ... find it hard to imagine as implemented unless we end up with multiple listeners executing on the consumer thread & a listener that precedes this one throws or something along those lines. And in that scenario right now I think we'd expect the exception to bubble out of KafkaConsumer.poll(), which would at least give us a clear signal that something went terribly wrong.

Reply:
We rely on `addListener` always resuling in listener being invoked in various places. So the code should be ok. But a comment that explains the sequence and exception scenarios would be useful.

Reply:
Ok I've added a comment describing the rationale behind the sequence. If this looks good, what are the next steps for getting this merged?